kube-core/src/object.rs

```diff
@@ -102,6 +102,46 @@ impl<'a, T: Clone> IntoIterator for &'a mut ObjectList<T> {
     }
 }
 
+/// A trait to access the `spec` of a Kubernetes resource.
+///
+/// Some built-in Kubernetes resources and all custom resources do have a `spec` field.
+/// This trait can be used to access this field.
+///
+/// This trait is automatically implemented by the kube-derive macro and is _not_ currently
+/// implemented for the Kubernetes API objects from `k8s_openapi`.
+///
+/// Note: Not all Kubernetes resources have a spec (e.g. `ConfigMap`, `Secret`, ...).
+pub trait HasSpec {
+    /// The type of the `spec` of this resource
+    type Spec;
+
+    /// Returns a reference to the `spec` object
+    fn spec(&self) -> &Self::Spec;
```

Comment:
Looks like the `spec` fields of most k8s-openapi types are also optional?

Reply:
That's a good point... I looked at a few core types and either they don't have a `spec` or it's optional.
I'm not 100% sure how it is with CRDs.

So...do we want to make this return an Option as well?

Reply:
The spec field being required here is a ergonomic choice on our end. We haven't seen any benefit to option wrapping it, and in general it's bad practice to encourage blind unwraps everywhere.

In general, I think it's only optional so that kubernetes can accept blank objects (which later fail validation, because they do some of the existence checks later within validation).

Reply:
IMO, I would say leave it as is. If we can find a benefit for making it not optional (such as incorporating into codegen output), then we can deal with that then (and even then i think we can just unwrap for users).

Reply:
Ideally it'd be required in `k8s-openapi` as well, and just initialized as `Self::Spec::default()`.

While `spec_mut` can do this sort of defaulting (since it works with a mutable ref), `spec` cannot update the original object (since it only gets a shared ref) and cannot return an independent default object (since it has to return a borrowed object). So the only option that remains there is to panic.

But it also just doesn't really sit right with me to panic on something like `Pod::default().spec()` (which seems like a code path that generic code would hit on pretty quickly).

Reply:
Okay, I'll leave it as is for now